network/framework/src/peer/mod.rs

```diff
@@ -461,21 +496,37 @@ where
                 );
             },
             NetworkMessage::RpcRequest(request) => {
-                if let Err(err) = self
-                    .inbound_rpcs
-                    .handle_inbound_request(&mut self.peer_notifs_tx, request)
-                {
-                    warn!(
-                        NetworkSchema::new(&self.network_context)
-                            .connection_metadata(&self.connection_metadata),
-                        error = %err,
-                        "{} Error handling inbound rpc request: {}",
-                        self.network_context,
-                        err
-                    );
+                match self.upstream_handlers.get(&request.protocol_id) {
+                    None => {
+                        counters::direct_send_messages(&self.network_context, UNKNOWN_LABEL).inc();
+                        counters::direct_send_bytes(&self.network_context, UNKNOWN_LABEL)
+                            .inc_by(request.raw_request.len() as u64);
+                    },
+                    Some(handler) => {
+                        let sender = self.connection_metadata.remote_peer_id;
+                        let network_id = self.network_context.network_id();
+                        let sender = PeerNetworkId::new(network_id, sender);
+                        if let Err(err) = self
+                            .inbound_rpcs
+                            .handle_inbound_request(handler, ReceivedMessage::new(message, sender))
+                        {
+                            warn!(
+                                NetworkSchema::new(&self.network_context)
+                                    .connection_metadata(&self.connection_metadata),
+                                error = %err,
+                                "{} Error handling inbound rpc request: {}",
+                                self.network_context,
+                                err
+                            );
+                        }
+                    },
                 }
             },
-            NetworkMessage::RpcResponse(response) => {
+            NetworkMessage::RpcResponse(_) => {
+                // non-reference cast identical to this match case
```

Comment:
What's the reason we can't just match this directly?

Reply:
the other branches work better with a match over a reference, and `*response` didn't work.